### PR TITLE
Lightning: parse parquet bit string

### LINKED ISSUE #37774: Lightning: parse bit string exported as Parquet from Aurora error
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
1. export table with schema from Aurora:
```
CREATE TABLE jxtest(
    `id` char(36) NOT NULL,
    `a` bigint unsigned NOT NULL,
    `aa` bigint signed NOT NULL,
    `b` int(11) unsigned NOT NULL,
    `bb` int(11) signed NOT NULL,
    `c` smallint signed NOT NULL,
    `cc` smallint unsigned NOT NULL,
    `d` tinyint signed NOT NULL,
    `dd` tinyint unsigned NOT NULL,
    `e` float unsigned NOT NULL,
    `ee` float signed NOT NULL,
    `f` VARCHAR(30) NOT NULL,
    `ff` TEXT NOT NULL,
    `h` MEDIUMTEXT NOT NULL,
    `hh` LONGTEXT NOT NULL,
    `ii` TINYTEXT NOT NULL,
    `j` DECIMAL NOT NULL,
    `jj` DECIMAL(8,0) NOT NULL,
    `k` DECIMAL(8,8) NOT NULL,
    `kk` DECIMAL(20,0) NOT NULL,
    `l` DECIMAL(20,8) NOT NULL,
    `ll` DECIMAL(36,0) NOT NULL,
    `m` DECIMAL(36,8) NOT NULL,
    `mm` DATE NOT NULL,
    `n` TIME NOT NULL,
    `nn` YEAR NOT NULL,
    `o` DATETIME NOT NULL,
    `oo` BINARY NOT NULL,
    `p` BLOB NOT NULL,
    `pp` LONGBLOB NOT NULL,
    `q` MEDIUMBLOB NOT NULL,
    `qq` TINYBLOB NOT NULL,
    `rr` BIT NOT NULL,
    `s` BOOLEAN NOT NULL,
    `ss` DOUBLE signed NOT NULL,
    `t` DOUBLE unsigned NOT NULL,
    PRIMARY KEY ( `id` ),
    KEY `index_a` (`a`) );
```
2. import parquet using lightning

### 2. What did you expect to see? (Required)
All data the same as that in Aurora
### 3. What did you see instead (Required)
diff:
for binary type,
In Aurora: b'111111111'
In TiDB: 0x31313131313131313131
### 4. What is your TiDB version? (Required)
TiDB: v6.2.0
Lighting: v6.2.0

### PR DESCRIPTION
**Pull request by buchuitoudegou**
### What problem does this PR solve?

Issue Number: close #37774 

Problem Summary:
Parse string without LogicalType.STRING as bit string.

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: br/pkg/lightning/mydump/parquet_parser.go [hunk 1]
```diff
@@ -453,6 +454,15 @@ func setDatumByString(d *types.Datum, v string, meta *parquet.SchemaElement) {
 	if meta.LogicalType != nil && meta.LogicalType.DECIMAL != nil {
 		v = binaryToDecimalStr([]byte(v), int(meta.LogicalType.DECIMAL.Scale))
 	}
+	if meta != nil && *meta.Type == parquet.Type_BYTE_ARRAY && meta.LogicalType == nil {
```
Comment (lance6716) [inline br/pkg/lightning/mydump/parquet_parser.go:457]: noticed there's `Type_FIXED_LEN_BYTE_ARRAY`, not sure if we should also convert it

File: br/pkg/lightning/mydump/parquet_parser.go [hunk 2]
```diff
@@ -453,6 +454,15 @@ func setDatumByString(d *types.Datum, v string, meta *parquet.SchemaElement) {
...
+		// for string that isn't logically string
+		// we should treat it as bit string.
+		// introduce to solve this issue: https://github.com/pingcap/tidb/issues/37774
+		ret, err := bitStringToByteSlice(v)
+		if err == nil {
+			v = string(ret)
```
Comment (lance6716) [inline br/pkg/lightning/mydump/parquet_parser.go:463]: maybe d.SetBytes

File: br/pkg/lightning/mydump/parquet_parser.go [hunk 3]
```diff
@@ -602,3 +612,28 @@ func int96ToTime(parquetDate []byte) time.Time {
...
+func bitStringToByteSlice(v string) ([]byte, error) {
+	ret := make([]byte, 0)
+	// traverse byte reversely
+	for i := len(v); i > 0; i -= 8 {
+		var str string
+		if i-8 < 0 {
+			str = v[0:i]
+			diff := 8 - len(str)
+			for j := 0; j < diff; j++ {
+				// padding with 0
```
Comment (lance6716) [inline br/pkg/lightning/mydump/parquet_parser.go:625]: no need to pad?

File: br/pkg/lightning/mydump/parquet_parser.go [hunk 4]
```diff
@@ -602,3 +612,28 @@ func int96ToTime(parquetDate []byte) time.Time {
...
+				str = "0" + str
+			}
+		} else {
+			str = v[i-8 : i]
+		}
+		b, err := strconv.ParseUint(str, 2, 8)
+		if err != nil {
+			return nil, err
+		}
+		// append to the result reversely
```
Comment (lance6716) [inline br/pkg/lightning/mydump/parquet_parser.go:635]: can you use more asymmetric and more various-length data in integration tests to check it's little/big endian

File: br/pkg/lightning/mydump/parquet_parser.go (hunk 1 above)
Comment (lance6716) [reply br/pkg/lightning/mydump/parquet_parser.go:457]: And how can we make sure we won't include other unwanted types by `*meta.Type == parquet.Type_BYTE_ARRAY && meta.LogicalType == nil`? Can you give some reference

Comment (buchuitoudegou) [reply br/pkg/lightning/mydump/parquet_parser.go:457]: > And how can we make sure we won't include other unwanted types by *meta.Type == parquet.Type_BYTE_ARRAY && meta.LogicalType == nil? Can you give some reference

It's more likely an expediency. In the manual of `parquet-go`, I didn't see how they handle bit string differently. 

Perhaps I should've read their code instead.

FYI:  https://pkg.go.dev/github.com/xitongsys/parquet-go#section-readme

Comment (buchuitoudegou) [reply br/pkg/lightning/mydump/parquet_parser.go:457]: > noticed there's Type_FIXED_LEN_BYTE_ARRAY, not sure if we should also convert it

According to the manual, `FIXED_LEN_BYTE_ARRAY ` is used to represent `DECIMAL` or `INTERVAL`.

**Pull request closed without merging**